File: src/main/java/cses/tree_algorithms/TreeDiameter.java
package cses.tree_algorithms;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class TreeDiameter {
    static ArrayList<Integer>[] adj;
    static int n;

    static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        adj = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) adj[i] = new ArrayList<>();

        for (int i = 0; i < n - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            adj[u].add(v);
            adj[v].add(u);
        }

        int endpoint = bfs(1)[0];
        int diameter = bfs(endpoint)[1];

        System.out.println(diameter);
    }

    // Returns {farthestNode, distance}
    static int[] bfs(int start) {
        boolean[] visited = new boolean[n + 1];
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{start, 0});
        visited[start] = true;

        int farthest = start;
        int maxDist = 0;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int u = cur[0];
            int dist = cur[1];

            if (dist > maxDist) {
                maxDist = dist;
                farthest = u;
            }

            for (int v : adj[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    queue.add(new int[]{v, dist + 1});
                }
            }
        }

        return new int[]{farthest, maxDist};
    }
}

File: src/main/kotlin/cses/tree_algorithms/TreeDiameter.kt
package cses.tree_algorithms

fun main() {
//    val n = readln().toInt()
//    val edges = mutableListOf<Pair<Int, Int>>()
//    for (i in 1..n) {
//        val (u, v) = readln().split(" ").map { it.toInt() }
//        edges.add(u to v)
//    }

    val n = 5
    val edges = mutableListOf<Pair<Int, Int>>(
        1 to 2,
        1 to 3,
        3 to 4,
        3 to 5
    )

    val answer = solution2(n, edges)

    println(answer)
}

private fun solution(n: Int, edges: List<Pair<Int, Int>>): Int {
    val adj = Array(n + 1) { mutableListOf<Int>() }
    for ((u, v) in edges) {
        adj[u].add(v)
        adj[v].add(u)
    }

    var visited = BooleanArray(n + 1)
    val stack = ArrayDeque<Pair<Int, Int>>()

    visited[1] = true

    adj[1].forEach { stack.add(it to 1) }

    var furthest = 0
    while (stack.isNotEmpty()) {
        val (u, parent) = stack.removeFirst()
        visited[u] = true
        furthest = u

        for (v in adj[u]) {
            if (v == parent) continue
            if (!visited[v]) stack.add(v to u)
        }
    }

    visited = BooleanArray(n + 1)
    visited[furthest] = true
    var work = adj[furthest].map { it to furthest }
    var distance = 0

    while (true) {
        if (work.isEmpty()) break
        distance++

        val newWork = mutableListOf<Pair<Int, Int>>()

        work.forEach { (u, parent) ->
            visited[u] = true

            for (v in adj[u]) {
                if (v == parent) continue
                if (!visited[v]) newWork.add(v to u)
            }
        }

        work = newWork
    }

    return distance
}

/**
 * ChatGpt polished version
 */
private fun solution2(n: Int, edges: List<Pair<Int, Int>>): Int {
    val adj = Array(n + 1) { mutableListOf<Int>() }
    for ((u, v) in edges) {
        adj[u].add(v)
        adj[v].add(u)
    }

    fun bfs(start: Int): Pair<Int, Int> {
        val visited = BooleanArray(n + 1)
        // I was really wanting to track distance in data. Shoulda trusted it
        val queue = ArrayDeque<Pair<Int, Int>>()

        // this is neat, better base case
        queue.add(start to 0)
        visited[start] = true

        var farthest = start
        var maxDist = 0

        while (queue.isNotEmpty()) {
            val (u, dist) = queue.removeFirst()
            if (dist > maxDist) {
                maxDist = dist
                farthest = u
            }
            for (v in adj[u]) {
                // visited is already tracking that parent
                if (!visited[v]) {
                    visited[v] = true
                    queue.add(v to dist + 1)
                }
            }
        }
        return farthest to maxDist
    }

    val (endpoint, _) = bfs(1)
    val (_, diameter) = bfs(endpoint)
    return diameter
}
